# $\alpha$=0.1

In [2]:
import torch
import numpy as np
import torchvision.transforms as transforms               
from torchvision.datasets import CIFAR10       
from torch.utils.data import DataLoader
import detectors
import timm
from src.temperature_scaling import ModelWithTemperature
from src.aps import aps_test

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# load fine-tuned model
model = timm.create_model("resnet18_cifar10", pretrained=True)
model.to(device)

# reprocess the images from CIFAR10
data_transform = transforms.Compose([
    transforms.ToTensor(),          # transfer to tensor
    transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])  # normalize
])
# load images from CIFAR10
dataset = CIFAR10(root="../../data", train=False, download=True, transform=data_transform)

# Temperature Scaling
temp_scal_loader = DataLoader(dataset, batch_size=32, shuffle=True)
model = ModelWithTemperature(model, temperature=5.0).to(device)
model.set_temperature(temp_scal_loader)

model.eval()

# The number of times the experiment is going to be repeated
num_runs = 10

# error rate
alpha = 0.1

aps_test(model, dataset, device, num_runs=10, alpha=0.1)

Files already downloaded and verified
Before temperature - NLL: 0.303, ECE: 0.040
Optimal temperature: 4.904
After temperature - NLL: 0.868, ECE: 0.465
APS Classification, Start!

Running experiment 1/10...
q_hat = 0.4826019108295441
Total set size: 6280
Total coverage sets: 4463
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.256
Average Coverage Rate in runs 1: 0.8926

Running experiment 2/10...
q_hat = 0.4829704314470292
Total set size: 6372
Total coverage sets: 4505
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.2744
Average Coverage Rate in runs 2: 0.901

Running experiment 3/10...
q_hat = 0.4765876710414887
Total set size: 6191
Total coverage sets: 4441
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.2382
Average Coverage Rate in runs 3: 0.8882

Running experiment 4/10...
q_hat = 0.4778558343648911
Total set size: 6150
Total coverage sets: 4451
Total samples amount: 5000
Average Prediction 

# $\alpha$=0.05

In [4]:
aps_test(model, dataset, device, num_runs=10, alpha=0.05)

APS Classification, Start!

Running experiment 1/10...
q_hat = 0.5306897461414338
Total set size: 7754
Total coverage sets: 4720
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.5508
Average Coverage Rate in runs 1: 0.944

Running experiment 2/10...
q_hat = 0.5301525145769119
Total set size: 7837
Total coverage sets: 4747
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.5674
Average Coverage Rate in runs 2: 0.9494

Running experiment 3/10...
q_hat = 0.528480499982834
Total set size: 7712
Total coverage sets: 4728
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.5424
Average Coverage Rate in runs 3: 0.9456

Running experiment 4/10...
q_hat = 0.5264313727617264
Total set size: 7595
Total coverage sets: 4718
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 1.519
Average Coverage Rate in runs 4: 0.9436

Running experiment 5/10...
q_hat = 0.52988540828228
Total set size: 7693
To

# Result
$\alpha$=0.1
- Final Average **Prediction Set Size： 1.26**
- Final Average **Coverage: 89.47%**  

$\alpha$=0.05
- Final Average **Prediction Set Size： 1.55**
- Final Average **Coverage: 94.63%**